### PR TITLE
Feat/stackerdb miners contract

### LINKED ISSUE #3931: Nakamoto[Block Proposal]: Miner constructs BlockProposals and stores them in StackerDB
In mockamoto, the `stacks-node` miner produces blocks directly, however, in the argon milestone, the miner needs to submit those blocks for consideration by the signer.

In order to do this, the miner must produce `BlockProposal` structs and store them to the StackerDB.

The miner also needs to watch the network for their prior `BlockProposal`'s acceptance before producing another proposal.

### LINKED ISSUE #4018: [Nakamoto] `.miners` boot contract for StackerDB-assisted block propagation to signers
Miners will need to propagate blocks to Stackers for signing.  To do this, there would be a dedicated boot contract called `.miners` which implements the StackerDB trait.  The miners would write their blocks to this StackerDB, and the Stacks node itself would update the contract state as part of processing a new burnchain block.

The StackerDB in `.miners` would have two slots: the current miner, and the parent (previous) miner.  The principals for these miners are derived from their `LeaderKeyRegisterOp`s' public key hashes (see SIP-021).  When the node processes a sortition, it allocates the new miner's `principal` a StackerDB slot by evicting the slot of the now-grandparent miner.  The parent miner retains their slot as-is.  In the event of no sortition, then the new miner's `principal` would simply be set to the parent miner's `principal`.

The parent miner and new miner write their blocks to their respective slots, and Stackers receive them by subscribing to the `.miners` StackerDB via their nodes' event listeners.  Stackers continue to sign the parent miner's blocks until they can produce a `TenureChange` transaction, at which point they cease to sign the parent miner's blocks and will only sign the new miner's blocks.

### PR DESCRIPTION
**Pull request by jcnelson**
Implements #4018, with the caveats listed in #4018 (namely, the `.miners` `StackerDBConfig` structure must be synthesized directly from chainstate due to a chicken-and-egg problem of needing `.miners` to reflect the latest sortition state before any blocks have been processed).

The code here is mostly complete and has some minimal unit tests.  The following features need to be added:

- [x] The Nakamoto node needs to be upgraded to write its proposed block to the `.miners` contract.  There's already a helper method in `NakamotoBlockBuilder` that will generate the chunk.

The following test coverage needs to be added:

- [x] A `TestPeer`-based unit test to verify that the `StackerDBConfig` for `.miners` is correctly synthesized after processing each Bitcoin block
* See test_make_miners_stackerdb_config

- [x] A `TestPeer`-based unit test to verify that two nodes can be instantiated, which both track `.miners`, and a `StackerDBChunkData` post to one will show up on the other (i.e. this shows that StackerDB can propagate `NakamotoBlock`s for miners)
* See miner_writes_proposed_block_to_stackerdb. Note this test does not have two nodes tracking miners but simply uses a stackerdb session to check that the block is found as expected in the appropriate slot. Does not bother to test stacker db propogation which is tested in other tests already. It also it not TestPeer based as the propose nakamoto block logic is found only within the Relayer for the neon node itself so really should be an integration test.

- [x] A Nakamoto integration test to ensure that the miner propagates the (unsigned) `NakamotoBlock` to the `.miners` contract
* See above point. 
   * Optional (can be a separate PR): test that the stacks signer receives the `NakamotoBlock`: will be done in another PR once stacks signer begins to process proposed blocks
   
As discussed earlier, I'm handing this off to @jferrant so I can focus on building out the Nakamoto p2p layer.

Closes https://github.com/stacks-network/stacks-core/issues/4018
Closes https://github.com/stacks-network/stacks-core/issues/3931

### CODE REVIEW

File: stackslib/src/chainstate/nakamoto/miner.rs
Comment (jferrant) [inline stackslib/src/chainstate/nakamoto/miner.rs:509]: Should this be creating a serialized NakamotoBlockProposal rather than just a serialized block?

Comment (jferrant) [reply stackslib/src/chainstate/nakamoto/miner.rs:509]: Got my answer! No, the miners should just propose a NakamotoBlock and the Stacks-Signer should fill in the remaining info for the "NakamotoBlockProposal" struct in the endpoint PR as their view of the chain is the "most valid" one (70% need to agree on their view of the chain state).

File: stackslib/src/net/chat.rs
Comment (kantai) [inline stackslib/src/net/chat.rs:1198]: This seems like it should be either (1) some kind of global, or (2) a state variable of `PeerNetwork` or `ConversationP2P`, rather than a function argument.

File: stackslib/src/net/p2p.rs
Comment (kantai) [inline stackslib/src/net/p2p.rs:5338]: In this error case, should the new config still be stored to `new_stackerdb_configs`?

File: stackslib/src/net/p2p.rs
Comment (kantai) [inline stackslib/src/net/p2p.rs:5336]:
```suggestion
                let new_config_res =
                    if stackerdb_contract_id == boot_code_id(MINERS_NAME, chainstate.mainnet) {
                        // .miners contract -- directly generate the config
                        NakamotoChainState::make_miners_stackerdb_config(sortdb)
                    } else {
                        // normal stackerdb contract
                        StackerDBConfig::from_smart_contract(
                            chainstate,
                            sortdb,
                            &stackerdb_contract_id,
                        )
                     };
                let new_config = new_config_res
                    .unwrap_or_else(|e| {
                                warn!(
                                    "Failed to load StackerDB config";
                                    "contract" => %stackerdb_contract_id,
                                    "err" => ?e,
                                );
                                StackerDBConfig::noop()
                     });
```

This suggestion changes the error behavior for the miners contract to use the noop db config. Is it necessary for `continue` to be used here? If so, why is it different in this regard than other stackerdb configs?

File: stackslib/src/chainstate/stacks/boot/miners.clar
Comment (kantai) [inline stackslib/src/chainstate/stacks/boot/miners.clar:21]: Why is it necessary for this contract to exist at all?

File: stackslib/src/chainstate/nakamoto/tests/mod.rs
Comment (kantai) [inline stackslib/src/chainstate/nakamoto/tests/mod.rs:1685]: This seems to create 10 miners, not 2

File: stackslib/src/chainstate/nakamoto/tests/mod.rs
Comment (kantai) [inline stackslib/src/chainstate/nakamoto/tests/mod.rs:1659]: Why is `3..13` being used? That seems like a very specific range, but I don't understand why.

File: stackslib/src/chainstate/nakamoto/tests/mod.rs
Comment (kantai) [inline stackslib/src/chainstate/nakamoto/tests/mod.rs:1706]: I don't understand these indexes. Why not `0..10`?

File: stackslib/src/chainstate/nakamoto/tests/mod.rs
Comment (kantai) [inline stackslib/src/chainstate/nakamoto/tests/mod.rs:1814]: What happens if `make_miners_stackerdb_config()` is invoked at `i == 3` ?

File: stackslib/src/chainstate/nakamoto/tests/mod.rs
Comment (kantai) [inline stackslib/src/chainstate/nakamoto/tests/mod.rs:1860]: Should the prior two miners still be able to create block proposals?

File: stackslib/src/chainstate/nakamoto/tests/mod.rs
Comment (kantai) [inline stackslib/src/chainstate/nakamoto/tests/mod.rs:1845]: Why is `write_count = i`?

File: stackslib/src/chainstate/nakamoto/tests/mod.rs
Comment (kantai) [inline stackslib/src/chainstate/nakamoto/tests/mod.rs:1824]: Should these configs be checked at the end of this test?

**Review by kantai — COMMENTED** [review]
This mostly LGTM, just a few comments and questions.

File: stackslib/src/chainstate/nakamoto/miner.rs
Comment (jferrant) [reply stackslib/src/chainstate/nakamoto/miner.rs]: Write count is the version of the chunk being written (must sequentially increase). i have updated this though in my commit such that it gets it from the database itself.

File: stackslib/src/chainstate/stacks/boot/miners.clar
Comment (jferrant) [reply stackslib/src/chainstate/stacks/boot/miners.clar:21]: I think it will werror out when "subscribing" to the stacker db instance if there is no contract found, but I can see if that is actually the case.

File: stackslib/src/net/p2p.rs
Comment (jferrant) [reply stackslib/src/net/p2p.rs:5336]: I actually wonder if this should error if it fails to make a MINERS stackerdb config...If we continue, we will have a more nebulous error down the line due to the missing .miners stacker db no? Otherwise, I think using a noop is a better option as well to just continuing.

File: stackslib/src/chainstate/stacks/boot/miners.clar
Comment (jcnelson) [reply stackslib/src/chainstate/stacks/boot/miners.clar:21]: The contract has to exist so the RPC endpoints will work.  Normally (i.e. for every other stacker DB), these methods are only evaluated every Bitcoin block in order to generate a (node-side) StackerDB configuration.  In the specific case of `.miners`, the node generates that configuration on its own.

Comment (kantai) [reply stackslib/src/chainstate/stacks/boot/miners.clar:21]: Does it make more sense to special case the RPC endpoints for `.miners`? The config lookups are already treated as special cases, so its not like this really avoids the need for special casing, and having a contract around that returns data that doesn't correspond to the actual behavior of the StackerDB seems confusing if it can be avoided.

File: stackslib/src/net/chat.rs
Comment (jcnelson) [reply stackslib/src/net/chat.rs:1198]: It gets passed in from the node implementation, since it is able to determine whether or not there are more blocks to process based on burnchain state.  I'd prefer to leave it this way.

File: stackslib/src/chainstate/stacks/boot/miners.clar
Comment (jcnelson) [reply stackslib/src/chainstate/stacks/boot/miners.clar:21]: We already expect users to understand that any contracts instantiated with the boot address behave differently than expected, so I don't think it's unreasonable to expect the same understanding for `.miners`.  Also, there are currently four StackerDB-specific endpoints, and some others (e.g. /v2/neighbors) that convey StackerDB data.  I'd rather not have to duplicate them all just for `.miners`.

File: stackslib/src/net/p2p.rs
Comment (jcnelson) [reply stackslib/src/net/p2p.rs:5336]: Ah, this is an oversight. If we fail to generate the `.miners` StackerDB config, then we should use the previous one instead of simply refusing to synthesize it.  Then at least the chain can continue.

This error path is reachable only from database errors in the SortitionDB, however.  If it fails, it may be best to just shut down the node since it's unclear how those errors could ever be recoverable.

File: stackslib/src/chainstate/nakamoto/tests/mod.rs
Comment (jcnelson) [reply stackslib/src/chainstate/nakamoto/tests/mod.rs:1689]: The comment is the bug.  There ought to be 10 miners.

File: stackslib/src/chainstate/nakamoto/tests/mod.rs
Comment (jcnelson) [reply stackslib/src/chainstate/nakamoto/tests/mod.rs:1661]: At some point there was special behavior for `i < 3`, but it no longer appears to be there.  It's fine to change this all to `0..10`, and remove any `i - 3`'s in the loops that follow (as well as any other place where `i` is compared to `3`).

File: stackslib/src/chainstate/stacks/boot/miners.clar
Comment (kantai) [reply stackslib/src/chainstate/stacks/boot/miners.clar:21]: Why would you need to duplicate the whole endpoints? What fails in those endpoints if the contract doesn't exist? It's true that boot address contracts do behave somewhat differently than normal contracts, but none of them actually return incorrect data (as this one would).

File: stackslib/src/chainstate/nakamoto/tests/mod.rs
Comment (jcnelson) [reply stackslib/src/chainstate/nakamoto/tests/mod.rs:1818]: It would panic because there would be no block-commit associated with the first-ever snapshot (which is what the code would return when asked for the last snapshot with a sortition, when there are no associated block-commits).

File: stackslib/src/chainstate/nakamoto/tests/mod.rs
Comment (jcnelson) [reply stackslib/src/chainstate/nakamoto/tests/mod.rs]: Why prior two?  I think we just want the prior

File: stackslib/src/chainstate/nakamoto/tests/mod.rs
Comment (jcnelson) [reply stackslib/src/chainstate/nakamoto/tests/mod.rs:1828]: Yes

File: stackslib/src/chainstate/nakamoto/tests/mod.rs
Comment (kantai) [reply stackslib/src/chainstate/nakamoto/tests/mod.rs:1818]: It seems like it would return an error of some kind from `get_last_snapshot_with_sortition` rather than panic. If that's the case, the test should check that error case.

File: stackslib/src/chainstate/stacks/boot/miners.clar
Comment (jcnelson) [reply stackslib/src/chainstate/stacks/boot/miners.clar:21]: > Why would you need to duplicate the whole endpoints?

It sounded like that's what you're asking for?  A set of API endpoints specific to querying, storing, and getting the metadata for chunks of data written for `.miners`?

> What fails in those endpoints if the contract doesn't exist?

The StackerDB API endpoints query the `stackerdbs.sqlite` database directly; they don't query the Clarity VM.  Instead, the `PeerNetwork` will do so whenever it detects that the burnchain has advanced -- it will re-evaluate these two Clarity functions to synthesize a `StackerDBConfig`, and will then update the `stackerdbs.sqlite` database with it (this is documented in `src/stackslib/net/stackerdb/mod.rs`).

If the contract doesn't exist -- i.e. if `PeerNetwork` doesn't synthesize a `StackerDBConfig` struct for its StackerDB replica -- then there won't be any configuration data for them in `stackerdbs.sqlite`, and thus the endpoint will fail.

> but none of them actually return incorrect data (as this one would).

The RPC endpoints don't query the smart contract for any state (as described above, they're only involved in the `StackerDBConfig` generation), so this shouldn't be an issue.

File: stackslib/src/net/p2p.rs
Comment (jferrant) [reply stackslib/src/net/p2p.rs]: I chose to have it error as I think this would be due to sortition DB errors which would indicate a much bigger problem

File: stackslib/src/chainstate/nakamoto/tests/mod.rs
Comment (jferrant) [reply stackslib/src/chainstate/nakamoto/tests/mod.rs]: I d obelieve this was to test the functionality in make_stackerdb_block_proposal was actually using a different version number each time based on the past in value. Its no longer relevant as I determine the version number on the fly now.

File: stackslib/src/clarity_vm/clarity.rs
Comment (jferrant) [reply stackslib/src/clarity_vm/clarity.rs:1307]: yes it is! Will use boot_code_tx_auth()

File: stackslib/src/net/chat.rs
Comment (jferrant) [reply stackslib/src/net/chat.rs:1198]: Do we have a consensus? Is PeerNetwork created from the node implementation? Cause in that case, making if a state variable of PeerNetwork would be what I would lean towards, but I am happy to do whatever.

File: stackslib/src/chainstate/nakamoto/tests/mod.rs
Comment (jferrant) [reply stackslib/src/chainstate/nakamoto/tests/mod.rs:1661]: Will do!

File: stackslib/src/chainstate/nakamoto/tests/mod.rs
Comment (jferrant) [reply stackslib/src/chainstate/nakamoto/tests/mod.rs:1689]: Will fix!

File: stackslib/src/chainstate/nakamoto/tests/mod.rs
Comment (jferrant) [reply stackslib/src/chainstate/nakamoto/tests/mod.rs:1710]: Will update to 0..10 :)

File: stackslib/src/chainstate/nakamoto/tests/mod.rs
Comment (jferrant) [reply stackslib/src/chainstate/nakamoto/tests/mod.rs:1828]: Will do!

File: stackslib/src/chainstate/nakamoto/tests/mod.rs
Comment (jferrant) [reply stackslib/src/chainstate/nakamoto/tests/mod.rs:1818]: It doesn't panic. But it does generate a result that I am having trouble figuring out what it means XD the assigned miner private key for the stacker dbs writers is not one I can seem to figure out where it comes from

File: stackslib/src/net/chat.rs
Comment (jcnelson) [reply stackslib/src/net/chat.rs:1198]: It needs to be an argument to `PeerNetwork::run()`, since `ibd` is computed from state that lives exclusively in the node implementation.  Please leave as-is.

**Review by jferrant — APPROVED** [review]

File: testnet/stacks-node/src/tests/nakamoto_integrations.rs
Comment (jcnelson) [inline testnet/stacks-node/src/tests/nakamoto_integrations.rs:151]: The constructor for `Config` needs to ensure that the `MINERS_NAME` contract is present in the list of StackerDBs if the node is mining.  It can't be opted out.  Same will go for `.stackers`.

File: testnet/stacks-node/src/tests/nakamoto_integrations.rs
Comment (jcnelson) [inline testnet/stacks-node/src/tests/nakamoto_integrations.rs:1028]: Is there any way you can verify that it's the same block mined?  E.g. does it have the expected transactions and so on?

File: testnet/stacks-node/src/nakamoto_node/miner.rs
Comment (jcnelson) [inline testnet/stacks-node/src/nakamoto_node/miner.rs:132]: I'm not sure this needs to be a member of `BlockMinerThread`.  You're basically connecting to yourself -- you can afford to instantiate a `StackerDBSession` each time you mine a block, because there's no gain to be had by preserving the TCP window across multiple block productions.

File: stackslib/src/net/p2p.rs
Comment (jcnelson) [reply stackslib/src/net/p2p.rs:5334]: Agreed.  Keep the old config on error.

File: stackslib/src/net/mod.rs
Comment (jcnelson) [inline stackslib/src/net/mod.rs:2139]: This shouldn't be necessary.  Instead, you can just call `peer.step_with_ibd(true)` in-between calls to `peer.next_burnchain_block()`. Then, the `.miners` contract DB state should initialize on its own.

File: stackslib/src/chainstate/nakamoto/coordinator/tests.rs
Comment (jcnelson) [inline stackslib/src/chainstate/nakamoto/coordinator/tests.rs:91]: Here, just call `peer.step_with_ibd(true)` after each call to `peer.tenure_with_txs(..)`.  That should initialize the `.miners` StackerDB.

**Review by jcnelson — COMMENTED** [review]
This overall LGTM. Thanks for taking it over @jferrant!  Just a couple comments need to be addressed and then I'll approve.

File: testnet/stacks-node/src/tests/nakamoto_integrations.rs
Comment (jferrant) [reply testnet/stacks-node/src/tests/nakamoto_integrations.rs:151]: kk will add!

File: testnet/stacks-node/src/tests/nakamoto_integrations.rs
Comment (jferrant) [reply testnet/stacks-node/src/tests/nakamoto_integrations.rs:1028]: I am sure there is. Will see if I can add it in :D

File: testnet/stacks-node/src/nakamoto_node/miner.rs
Comment (jferrant) [reply testnet/stacks-node/src/nakamoto_node/miner.rs:132]: Alright will do :D

File: stackslib/src/net/mod.rs
Comment (jferrant) [reply stackslib/src/net/mod.rs:2139]: Ah kk ! I was wondering how ot enforce the stacker db miners being instantiated and this seemed the least janky option (though still janky). Will try with peer.step_with_ibd and see what happens!

File: stackslib/src/chainstate/nakamoto/coordinator/tests.rs
Comment (jferrant) [reply stackslib/src/chainstate/nakamoto/coordinator/tests.rs:91]: Hm it still breaks the make_stackerdb_config test. But perhaps I should just change that test D:

File: stackslib/src/net/mod.rs
Comment (jferrant) [reply stackslib/src/net/mod.rs:2139]: Actually I remember why I did this. Because it breaks test_make_miners_stackerdb_config as it attempts to build a block proposal which makes the config using the stacker db .miners contract which doesn't exist. Not sure how to fix this.

File: testnet/stacks-node/src/tests/nakamoto_integrations.rs
Comment (jferrant) [reply testnet/stacks-node/src/tests/nakamoto_integrations.rs:1029]: Done!

File: testnet/stacks-node/src/nakamoto_node/miner.rs
Comment (jferrant) [reply testnet/stacks-node/src/nakamoto_node/miner.rs]: Done.

File: testnet/stacks-node/src/tests/nakamoto_integrations.rs
Comment (jferrant) [reply testnet/stacks-node/src/tests/nakamoto_integrations.rs:176]: I am not sure which contructor you are referencing? do you mean the default one or the one which is created from a ConfigFile? I feel this might be better ina. sep PR.

File: stackslib/src/chainstate/stacks/boot/miners.clar
Comment (kantai) [reply stackslib/src/chainstate/stacks/boot/miners.clar:21]: > It sounded like that's what you're asking for? A set of API endpoints specific to querying, storing, and getting the metadata for chunks of data written for .miners?

I see -- no, that's not what I was suggesting. I was suggesting adding a check anywhere that would fail without the presence of a contract for `.miners`.

> The RPC endpoints don't query the smart contract for any state (as described above, they're only involved in the StackerDBConfig generation), so this shouldn't be an issue.

Yep, I see that now, which still leaves me wondering why the `.miners` contract needs to exist.

> If the contract doesn't exist -- i.e. if PeerNetwork doesn't synthesize a StackerDBConfig struct for its StackerDB replica -- then there won't be any configuration data for them in stackerdbs.sqlite, and thus the endpoint will fail.

But this is exactly what this PR _avoids_ for `.miners`, right? `create_or_reconfigure_stackerdbs` has a special case for `.miners`, and never actually invokes the Clarity VM. So, it doesn't seem to need the contract to exist either.

Applying this diff to the PR, for example, still leaves the end-to-end test passing -- meaning that the stackerdbs instance still initializes the miners stackerdb, still accepts the block proposal, and still returns it, all without `.miners` ever existing. If its actually the case that the contract doesn't need to exist for the stackerdb to work, I'd much prefer not instantiating it at all. As you say, the contract doesn't respond on those public functions with data that actually corresponds to anything.

```diff
diff --git a/stackslib/src/clarity_vm/clarity.rs b/stackslib/src/clarity_vm/clarity.rs
index 9a18decfd..8bbd1cb95 100644
--- a/stackslib/src/clarity_vm/clarity.rs
+++ b/stackslib/src/clarity_vm/clarity.rs
@@ -1288,28 +1288,6 @@ impl<'a, 'b> ClarityBlockConnection<'a, 'b> {
             let miners_contract_tx =
                 StacksTransaction::new(tx_version.clone(), boot_code_auth.clone(), payload);
 
-            let miners_initialization_receipt = self.as_transaction(|tx_conn| {
-                // initialize with a synthetic transaction
-                debug!("Instantiate {} contract", &miners_contract_id);
-                let receipt = StacksChainState::process_transaction_payload(
-                    tx_conn,
-                    &miners_contract_tx,
-                    &boot_code_account,
-                    ASTRules::PrecheckSize,
-                )
-                .expect("FATAL: Failed to process .miners contract initialization");
-                receipt
-            });
-
-            if miners_initialization_receipt.result != Value::okay_true()
-                || miners_initialization_receipt.post_condition_aborted
-            {
-                panic!(
-                    "FATAL: Failure processing .miners contract initialization: {:#?}",
-                    &miners_initialization_receipt
-                );
-            }
-
             /////////////////// .pox-4 ////////////////////////
             let first_block_height = self.burn_state_db.get_burn_start_height();
             let pox_prepare_length = self.burn_state_db.get_pox_prepare_length();
```

File: stackslib/src/chainstate/nakamoto/mod.rs
Comment (jcnelson) [inline stackslib/src/chainstate/nakamoto/mod.rs:3058]: Can this be an argument, instead? This is called from the network main loop, so ideally we don't do any gratuitous I/O.

Comment (jferrant) [reply stackslib/src/chainstate/nakamoto/mod.rs:3058]: Can do!

File: testnet/stacks-node/src/config.rs
Comment (jferrant) [inline testnet/stacks-node/src/config.rs:2223]: This was actually a bugfix. It used to have a double negation.

File: testnet/stacks-node/src/config.rs
Comment (jferrant) [inline testnet/stacks-node/src/config.rs:938]: I assumed it was better to forcibly add subscription to the stacker db miners contract rather than to error out. But let me know if this is not the case.

**Review by kantai — COMMENTED** [review]
Overall, LGTM, I just have one unresolved comment about the `.miners` contract. It seems like it doesn't actually need to exist, and would prefer to just remove it in that case.

File: stackslib/src/chainstate/stacks/boot/miners.clar
Comment (jferrant) [reply stackslib/src/chainstate/stacks/boot/miners.clar:21]: @jcnelson if nothing breaks to remove the .miners contract, I am happy to do so. I thought you mentioned that the RPC calls would fail though D:

**Comment by jcnelson** [comment]
Well, since I'm the one who opened it, I can't approve it, but LGTM.  Thanks for carrying this over the finish line @jferrant!  Just a couple comments to address before merging :pray:

File: stackslib/src/chainstate/stacks/boot/miners.clar
Comment (jcnelson) [reply stackslib/src/chainstate/stacks/boot/miners.clar:21]: I thought they would, but it turns out I was wrong -- the RPC endpoint only looks at the `stackerdbs.sqlite` DB to service queries, not the smart contract in the chainstate.  So as long as the `StackerDBConfig` is getting synthesized and applied, then the RPC endpoints should work.

**Review by kantai — APPROVED** [review]
LGTM, appreciate you for getting this across the line @jferrant!

**Pull request merged**
